File: Java/56.merge-intervals.java
/*
 * @lc app=leetcode id=56 lang=java
 *
 * [56] Merge Intervals
 */

// @lc code=start
import java.util.*;
class Solution {
    /*
    利用一个pointer cur记录最大结束时间，并进行更新，如果当前没有interval就把cur加到结果数组中
    注意： 最后记得把最后一个cur加到结果数组中
    time: O(nlogn)
    space: O(n)
    */
    public int[][] merge(int[][] intervals) {
        if (intervals == null || intervals.length <= 1) return intervals;
        Arrays.sort(intervals, (a, b) -> (a[0] - b[0]));
        int[] cur = intervals[0];
        List<int[]> res = new ArrayList<>();
        for (int i = 0; i < intervals.length; ++i) {
            if (cur[1] >= intervals[i][0]) {
                cur[1] = Math.max(cur[1], intervals[i][1]);
            } else {
                res.add(cur);
                cur = intervals[i];
            }
        }
        res.add(cur);
        return res.toArray(new int[res.size()][]);
    }
}
// @lc code=end

File: Cpp/56.merge-intervals.cpp
/*
 * @lc app=leetcode id=56 lang=cpp
 *
 * [56] Merge Intervals
 */

// @lc code=start
class Solution {
public:
    static bool comparator(vector<int>& v1, vector<int>& v2) {
        return v1[0] < v2[0];
    }
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.empty() || intervals.size() < 2) return intervals;
        vector<vector<int>> res;
        sort(intervals.begin(), intervals.end(), comparator);
        res.push_back(intervals[0]);
        for (int i = 1; i < intervals.size(); ++i) {
            if (intervals[i][0] <= res.back()[1]) res.back()[1] = max(res.back()[1], intervals[i][1]);
            else res.push_back(intervals[i]);
        }
        return res;
    }
};
// @lc code=end

